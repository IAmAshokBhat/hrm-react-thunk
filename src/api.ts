import {
  ILoginData,
  INewAttendance,
  INewLeave,
  IRegisterAPI
} from './constants';
import { fetchWithToken, getToken } from './utils';

const headers = {
  'X-JWT-Token': getToken()
};

export const getUserListAPI = () => fetchWithToken('user/list');

export const getUserDetailsAPI = (id: number) =>
  fetchWithToken(`api/users/${id}`);

export const getAttendanceDetailsAPI = (id: number, from: string, to: string) =>
  fetchWithToken(`api/attendance/${id}?from=${from}&to=${to}`);

export const getLeaveBalanceAPI = (id: number) =>
  fetchWithToken(`api/leaves/${id}`);

export const getEventDetailsAPI = (id: number, from: string, to: string) =>
  fetchWithToken(`api/events/${id}?from=${from}&to=${to}`);

export const getAllLeavesAPI = (id: number) =>
  fetchWithToken(`api/all-leaves/${id}`);

export const applyLeaveAPI = (newLeave: INewLeave) =>
  fetchWithToken(`api/apply-leave`, {
    method: 'POST',
    body: JSON.stringify(newLeave)
  });

export const addAttendanceAPI = (newAttendance: INewAttendance) =>
  fetchWithToken(`api/attendance`, {
    method: 'POST',
    body: JSON.stringify(newAttendance)
  });

export const loginAPI = (userInfo: ILoginData) =>
  fetchWithToken(`login`, {
    method: 'POST',
    body: JSON.stringify(userInfo)
  });

export const getDesignationListAPI = () =>
  fetchWithToken(`designation`, { headers: headers });

export const registerAPI = (newUser: IRegisterAPI) =>
  fetchWithToken(`user`, {
    method: 'POST',
    body: JSON.stringify(newUser)
  });
