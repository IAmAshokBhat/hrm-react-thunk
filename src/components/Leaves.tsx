import Card from '@mui/material/Card';
import Grid from '@mui/material/Grid';
import Paper from '@mui/material/Paper';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Bar, BarChart, Legend, Tooltip, XAxis, YAxis } from 'recharts';
import {
  fetchAllLeavesAction,
  fetchLeaveBalanceAction
} from '../redux/actions';
import {
  selectAllLeaves,
  selectLeaveBalance,
  selectLeaveCountsByMonth
} from '../redux/selectors';
import { getUserId } from '../utils';

export const Leaves = () => {
  const dispatch = useDispatch();
  const loggedInUser = getUserId();
  const holidayTaken = useSelector(selectLeaveCountsByMonth);
  const leaveData = useSelector(selectAllLeaves);
  const leaveBalance = useSelector(selectLeaveBalance);

  useEffect(() => {
    dispatch(fetchAllLeavesAction(loggedInUser));
    dispatch(fetchLeaveBalanceAction(loggedInUser));
  }, [dispatch, loggedInUser]);

  return (
    <Grid container spacing={2}>
      <Grid item md={12}>
        <Card>
          <BarChart width={1500} height={350} data={holidayTaken} barSize={30}>
            <XAxis dataKey="month" />
            <YAxis />
            <Tooltip />
            <Legend />
            <Bar dataKey="days" fill="#2976d2" />
          </BarChart>
        </Card>
      </Grid>
      <Grid item md={8}>
        <Card variant="outlined">
          <TableContainer component={Paper}>
            <Table sx={{ minWidth: 490 }} aria-label="Leave details">
              <TableHead>
                <TableRow>
                  <TableCell>Leave Type</TableCell>
                  <TableCell>No. Of Days</TableCell>
                  <TableCell>From</TableCell>
                  <TableCell>To</TableCell>
                  <TableCell>Status</TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                {leaveData.map(({ leaveType, from, to, status, noOfDays }) => (
                  <TableRow
                    key={from.toString()}
                    sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
                  >
                    <TableCell component="th" scope="row">
                      {leaveType}
                    </TableCell>
                    <TableCell component="th" scope="row">
                      {noOfDays}
                    </TableCell>
                    <TableCell component="th" scope="row">
                      {from}
                    </TableCell>
                    <TableCell component="th" scope="row">
                      {to}
                    </TableCell>
                    <TableCell component="th" scope="row">
                      {status}
                    </TableCell>
                  </TableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        </Card>
      </Grid>
      <Grid item md={4}>
        <Card variant="outlined">
          <TableContainer component={Paper}>
            <Table
              sx={{ minWidth: 490 }}
              aria-label="Leave balance"
              size="small"
            >
              <TableHead>
                <TableRow>
                  <TableCell>Leave Type</TableCell>
                  <TableCell>Balance</TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                {leaveBalance.map(({ leaveType, balance }) => (
                  <TableRow
                    key={leaveType}
                    sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
                  >
                    <TableCell component="th" scope="row">
                      {leaveType}
                    </TableCell>
                    <TableCell component="th" scope="row">
                      {balance}
                    </TableCell>
                  </TableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        </Card>
      </Grid>
    </Grid>
  );
};
