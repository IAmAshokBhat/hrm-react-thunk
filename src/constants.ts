export interface ILeave {
  leaveType: string;
  from: Date;
  to: Date;
  status: string;
  noOfDays: number;
}

export interface ILeaveInfo {
  from: string;
  to: string;
  category: string;
}

export interface ISelectLeaveType {
  value: number;
  label: string;
  balance: number;
}

export enum ACTIONS {
  GET_USERS = 'GET_USERS',
  GET_USERS_LOADING = 'GET_USERS_LOADING',
  GET_USERS_SUCCESS = 'GET_USERS_SUCCESS',
  GET_USERS_FAILURE = 'GET_USERS_FAILURE',

  GET_USER_DETAILS = 'GET_USER_DETAILS',
  GET_USER_DETAILS_LOADING = 'GET_USER_DETAILS_LOADING',
  GET_USER_DETAILS_SUCCESS = 'GET_USER_DETAILS_SUCCESS',
  GET_USER_DETAILS_FAILURE = 'GET_USER_DETAILS_FAILURE',

  GET_ATTENDANCE_DETAILS = 'GET_ATTENDANCE_DETAILS',
  GET_ATTENDANCE_DETAILS_LOADING = 'GET_ATTENDANCE_DETAILS_LOADING',
  GET_ATTENDANCE_DETAILS_SUCCESS = 'GET_ATTENDANCE_DETAILS_SUCCESS',
  GET_ATTENDANCE_DETAILS_FAILURE = 'GET_ATTENDANCE_DETAILS_FAILURE',

  GET_LEAVE_BALANCE = 'GET_LEAVE_BALANCE',
  GET_LEAVE_BALANCE_LOADING = 'GET_LEAVE_BALANCE_LOADING',
  GET_LEAVE_BALANCE_SUCCESS = 'GET_LEAVE_BALANCE_SUCCESS',
  GET_LEAVE_BALANCE_FAILURE = 'GET_LEAVE_BALANCE_FAILURE',

  GET_EVENT_DETAILS = 'GET_EVENT_DETAILS',
  GET_EVENT_DETAILS_LOADING = 'GET_EVENT_DETAILS_LOADING',
  GET_EVENT_DETAILS_SUCCESS = 'GET_EVENT_DETAILS_SUCCESS',
  GET_EVENT_DETAILS_FAILURE = 'GET_EVENT_DETAILS_FAILURE',

  GET_ALL_LEAVES = 'GET_ALL_LEAVES',
  GET_ALL_LEAVES_LOADING = 'GET_ALL_LEAVES_LOADING',
  GET_ALL_LEAVES_SUCCESS = 'GET_ALL_LEAVES_SUCCESS',
  GET_ALL_LEAVES_FAILURE = 'GET_ALL_LEAVES_FAILURE',

  GET_APPLY_LEAVE = 'GET_APPLY_LEAVE',
  GET_APPLY_LEAVE_LOADING = 'GET_APPLY_LEAVE_LOADING',
  GET_APPLY_LEAVE_SUCCESS = 'GET_APPLY_LEAVE_SUCCESS',
  GET_APPLY_LEAVE_FAILURE = 'GET_APPLY_LEAVE_FAILURE',

  GET_ADD_ATTENDANCE = 'GET_ADD_ATTENDANCE',
  GET_ADD_ATTENDANCE_LOADING = 'GET_ADD_ATTENDANCE_LOADING',
  GET_ADD_ATTENDANCE_SUCCESS = 'GET_ADD_ATTENDANCE_SUCCESS',
  GET_ADD_ATTENDANCE_FAILURE = 'GET_ADD_ATTENDANCE_FAILURE',

  GET_LOGIN = 'GET_LOGIN',
  LOGIN_LOADING = 'GET_LOGIN_LOADING',
  LOGIN_SUCCESS = 'GET_LOGIN_SUCCESS',
  LOGIN_FAILURE = 'GET_LOGIN_FAILURE',

  GET_DESIGNATION_LIST = 'GET_DESIGNATION_LIST',
  GET_DESIGNATION_LIST_LOADING = 'GET_DESIGNATION_LIST_LOADING',
  GET_DESIGNATION_LIST_SUCCESS = 'GET_DESIGNATION_LIST_SUCCESS',
  GET_DESIGNATION_LIST_FAILURE = 'GET_DESIGNATION_LIST_FAILURE',

  REGISTER = 'REGISTER',
  REGISTER_LOADING = 'REGISTER_LOADING',
  REGISTER_SUCCESS = 'REGISTER_SUCCESS',
  REGISTER_FAILURE = 'REGISTER_FAILURE'
}

export interface IUserAPI {
  user_id: number;
  name: string;
}
export interface IUserListAPI {
  data: IUserAPI[];
}

export interface IUserInfoDataAPI {
  user_id: number;
  name: string;
  designation: string;
  manager: string;
  email: string;
}

export interface IUserInfoAPI {
  data: IUserInfoDataAPI[];
}

export interface IUserInfo {
  userId: number;
  name: string;
  designation: string;
  manager: string;
  email: string;
}

interface ILeaveBalanceItemAPI {
  leave_type_id: number;
  leave_type: string;
  balance: number;
}

export interface ILeaveBalanceAPI {
  data: ILeaveBalanceItemAPI[];
}

export interface ILeaveBalance {
  leaveTypeId: number;
  leaveType: string;
  balance: number;
}

interface IAttendanceItemAPI {
  name: string;
  attendance_date: string;
  start_time: string;
  end_time: string;
}

export interface IAttendanceDetailsAPI {
  data: IAttendanceItemAPI[];
}

export interface IAttendanceItem {
  name: string;
  attendanceDate: string;
  startTime: string;
  endTime: string;
  noOfHours: number;
}

export interface IError {
  ok: boolean;
  message: string;
  status: number;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
interface IAsyncState<T = any> {
  error: IError | null;
  loading: boolean;
  value: T;
}

export interface IDefaultRoot {
  userList: IAsyncState<ILabelValue[]>;
  userInfo: IAsyncState<IUserInfo | null>;
  leaveBalance: IAsyncState<ILeaveBalance[]>;
  attendanceDetails: IAsyncState<IAttendanceItem[]>;
  evetDetails: IAsyncState<IEvent[]>;
  allLeaves: IAsyncState<ILeave[]>;
  leaveCountByMonth: IAsyncState<ILeaveCountByMonth[] | []>;
  applyLeave: IAsyncState<IGenericResponse>;
  addAttendance: IAsyncState<IGenericResponse>;
  loginDetails: IAsyncState<ILoginAPI>;
  designationList: IAsyncState<ILabelValue[]>;
  registeration: IAsyncState<IGenericResponse>;
}

export interface IAction {
  type: string;
}

export interface IUserListAction extends IAction {
  payload: IUserAPI[] | null;
}

export interface IUserInfoAction extends IAction {
  payload: IUserInfoDataAPI | null;
}

export interface ILeaveBalanceAction extends IAction {
  payload: ILeaveBalanceItemAPI[] | null;
}

export interface IAttendanceDetailsAction extends IAction {
  payload: IAttendanceItemAPI[] | null;
}

export interface IEventDetailsAction extends IAction {
  payload: IEventItemAPI[] | null;
}

interface ILeaveItemAPI {
  from_date: string;
  to_date: string;
  leave_type: string;
  status: number;
}

export interface IAllLeavesAction extends IAction {
  payload: ILeaveItemAPI[] | null;
}

export interface IAllLeavesAPI {
  data: ILeaveItemAPI[];
}

//TODO
export type TFAttendanceDetailsAction = (
  id: number,
  from: string,
  to: string
) => Promise<Response>;

interface IEventItemAPI {
  attendance_date?: string;
  start_time?: string;
  end_time?: string;
  from_date?: string;
  to_date?: string;
  leave_type?: string;
}

export interface IEventDetailsAPI {
  data: IEventItemAPI[];
}

export const DATE_FORMAT = 'DD-MM-YYYY';
export const DATE_FORMAT_CALENDAR = 'YYYY-MM-DD';

export enum EventType {
  WORKING = 'WORKING',
  LEAVE = 'LEAVE'
}

export interface IEvent {
  title: string;
  start: string;
  end: string;
  allDay?: boolean;
}

export interface IKeyValue {
  [key: string]: number;
}

export interface ILeaveCountByMonth {
  days: number;
  month: string;
}

export interface INewLeave {
  leaveTypeId: number;
  fromDate: string;
  toDate: string;
  userId: number;
}

export interface IGenericResponse<T = unknown> {
  status: number;
  message: string;
  data: T;
}

export interface IGenericAction extends IAction {
  payload: IGenericResponse | null;
}

export const enum NotificationType {
  SUCCESS = 'success',
  ERROR = 'error',
  WARNING = 'warning',
  INFO = 'info'
}

export interface INewAttendance {
  userId: number;
  date: string;
  inTime: string;
  outTime: string;
}

export interface ILoginData {
  userId: number;
  password: string;
}

export interface ILoginAPI {
  token: string;
  issued: number;
  expires: number;
  loginStatus: number;
  message: string;
  userId: number;
}

export interface ILoginAction extends IAction {
  payload: ILoginAPI | null;
}

export interface IRegistration {
  name: string;
  password: string;
  confirmPassword: string;
  email: string;
}

export interface ILabelValue {
  value: number;
  label: string;
}

export interface IDesignationListAPI {
  data: ILabelValue[];
}

export interface IDesignationListAction extends IAction {
  payload: ILabelValue[] | null;
}

export interface IRegisterAPI {
  name: string;
  designation: number;
  manager: number;
  password: string;
  email: string;
}

export interface IRegisterAction extends IAction {
  payload: IRegisterAPI | null;
}

export type ExpirationStatus = 'expired' | 'active' | 'grace';

export interface Session {
  id: number;
  dateCreated: number;
  username: string;
  issued: number;
  expires: number;
}

export type DecodeResult =
  | {
      type: 'valid';
      session: Session;
    }
  | {
      type: 'integrity-error';
    }
  | {
      type: 'invalid-token';
    };

export type TAnyObject = { [key: string]: any }; // eslint-disable-line @typescript-eslint/no-explicit-any

export type TAny = any; // eslint-disable-line @typescript-eslint/no-explicit-any
