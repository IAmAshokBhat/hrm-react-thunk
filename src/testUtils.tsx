import { render, RenderOptions, waitFor } from '@testing-library/react';
import { SnackbarProvider } from 'notistack';
import { ReactElement } from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { applyMiddleware, createStore, Store } from 'redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { IDefaultRoot, TAny, TAnyObject } from './constants';
import { TFAction } from './contracts';
import { reducers } from './redux/reducers';
import {
  defaultArrayState,
  defaultNullState,
  defaultState
} from './redux/reducers/__tests__/fixtures';

export const middlewares = [thunk];

export const createStoreWithMiddlewares = (
  initialState: TAnyObject = {}
): Store => {
  return createStore(reducers, initialState, applyMiddleware(...middlewares));
};
interface IRenderWithProviderOptions {
  initialStoreState?: TAnyObject;
  renderFunction?: TAny;
  options?: RenderOptions;
  route?: string;
  path?: string;
}

export const renderWithProvider = (
  ui: ReactElement,
  renderOptions?: IRenderWithProviderOptions
) => {
  const {
    initialStoreState = {},
    renderFunction = render,
    options = {},
    route = '',
    path = ''
  } = renderOptions || {};

  const appStore = createStoreWithMiddlewares(initialStoreState);

  let wrappedUi = (
    <SnackbarProvider
      anchorOrigin={{
        vertical: 'top',
        horizontal: 'right'
      }}
    >
      <Provider store={appStore}>{ui}</Provider>
    </SnackbarProvider>
  );

  if (route) {
    window.history.pushState({}, '', route);

    wrappedUi = (
      <Provider store={appStore}>
        <BrowserRouter>
          <Routes>
            <Route path={path} element={ui} />
          </Routes>
        </BrowserRouter>
      </Provider>
    );
  }

  return renderFunction(wrappedUi, options);
};

type TFApiCall = (...params: TAny) => Promise<Response>;

export interface IResponseMock {
  headers?: object;
  ok?: boolean;
  status?: number;
  statusText?: string;
  type?: ResponseType;
  url?: string;
  body?: TAny | null;
  bodyUsed?: boolean;
  blob?: () => Promise<TAny>;
}

const dummyPromise = () => new Promise((resolve) => resolve('dummy'));

export class ResponseMock {
  headers: object;

  ok: boolean;

  status: number;

  statusText: string;

  type: ResponseType;

  url: string;

  body: ReadableStreamDefaultReadResult<Uint8Array> | null;

  bodyUsed: boolean;

  blob: () => Promise<TAny>;

  constructor(props: IResponseMock = {}) {
    this.headers = props.headers || {};
    this.status = props.status || 200;
    this.ok = props.ok || this.status === 200;
    this.statusText = props.statusText || 'OK';
    this.type = props.type || 'basic';
    this.url = props.url || 'http://localhost';
    this.body = props.body;
    this.bodyUsed = props.bodyUsed || this.body !== null;
    this.blob = props.blob || dummyPromise;
  }

  json: () => Promise<TAny> = () => {
    const body = this.body || '""';

    try {
      return Promise.resolve(JSON.parse(body as string));
    } catch {
      return Promise.resolve(body);
    }
  };

  text: () => Promise<string> = () => Promise.resolve(this.statusText);
}

export const mockAPICall = (func: TFApiCall, response: IResponseMock) =>
  (func as jest.Mock).mockResolvedValue(new ResponseMock(response));

export const generateActionTests = async (
  actionName: string,
  action: TFAction,
  api: TAny,
  actionPayload: TAny = []
) => {
  const response = new ResponseMock({ ok: true, body: { data: [] } });

  const mockStore = configureStore(middlewares);

  (api as jest.Mock).mockReturnValue(Promise.resolve(response));
  const store = mockStore({});

  //TODO: Fix TS error
  // @ts-ignore
  await store.dispatch(action);

  await waitFor(() => {
    expect(store.getActions()).toEqual([
      {
        type: `${actionName}_LOADING`
      },
      {
        type: `${actionName}_SUCCESS`,
        //TODO: Accept action params from function params
        payload: actionPayload
      }
    ]);
  });
};

export const generateReducerTests = (
  reducerName: string,
  reducer: TAny,
  selector: TAny,
  actionName: string,
  successValue?: TAny,
  defaultValue?: TAny,
  expectedValue?: TAny,
  reducerLabel?: string
) => {
  describe(`for ${reducerName}`, () => {
    const reducerLbl = reducerLabel
      ? reducerLabel
      : reducerName.replace('Reducer', '');
    it(`${reducerName} reducer loading`, () => {
      reducer(undefined, {
        type: `${actionName}_LOADING`,
        payload: null
      });
      // TODO: Check why its failing
      // expect(selector(defaultState).loading).toEqual(true);
      expect(selector(defaultState).value).toEqual(
        defaultValue === null ? defaultValue : []
      );
      expect(selector(defaultState).error).toEqual(null);
    });

    it(`${reducerName} reducer success`, () => {
      const successState: IDefaultRoot = {
        ...defaultState,
        [reducerLbl]: {
          ...defaultNullState,
          value: expectedValue ? expectedValue : successValue
        }
      };
      reducer(defaultArrayState, {
        type: `${actionName}_SUCCESS`,
        payload: successValue
      });

      expect(selector(successState).error).toEqual(null);
      expect(selector(successState).loading).toEqual(false);
      expect(selector(successState).value).toEqual(
        expectedValue ? expectedValue : successValue
      );
    });

    it(`${reducerName} reducer error`, () => {
      const error = { message: '', status: 1, ok: false };
      const failedState: IDefaultRoot = {
        ...defaultState,
        [reducerLbl]: {
          ...defaultNullState,
          value: [],
          error
        }
      };

      reducer(undefined, {
        type: `${actionName}_FAILURE`,
        payload: error
      });

      expect(selector(failedState).loading).toEqual(false);
      expect(selector(failedState).value).toEqual([]);
      expect(selector(failedState).error).toEqual(error);
    });
  });
};
