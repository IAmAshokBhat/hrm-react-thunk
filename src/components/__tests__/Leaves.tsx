import { getAllLeavesAPI, getLeaveBalanceAPI } from '../../api';
import { TAnyObject } from '../../constants';
import { mockAPICall, renderWithProvider } from '../../testUtils';
import { Leaves } from '../Leaves';
import {
  allLeavesAPIResponse,
  defaultState,
  leaveBalanceAPIResponse
} from './fixtures';

jest.mock('../../api', () => ({
  getAllLeavesAPI: jest.fn(),
  getLeaveBalanceAPI: jest.fn()
}));

const LeaveBalanceResponse = {
  ok: true,
  body: leaveBalanceAPIResponse
};

const allLeavesResponse = {
  ok: true,
  body: allLeavesAPIResponse
};

describe('Leaves Page', () => {
  const generateNode = (initialStoreState: TAnyObject = defaultState) =>
    renderWithProvider(<Leaves />, {
      initialStoreState
    });

  beforeEach(() => {
    mockAPICall(getAllLeavesAPI, allLeavesResponse);
    mockAPICall(getLeaveBalanceAPI, LeaveBalanceResponse);
  });

  it('should render and match snapshot', () => {
    const { container } = generateNode();

    expect(container).toMatchSnapshot();
  });
});
