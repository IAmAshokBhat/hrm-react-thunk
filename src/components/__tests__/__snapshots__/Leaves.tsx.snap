// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaves Page should render and match snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-12 css-hagxv6-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
      >
        <div
          class="recharts-wrapper"
          style="position: relative; cursor: default; width: 1500px; height: 350px;"
        >
          <svg
            class="recharts-surface"
            height="350"
            version="1.1"
            viewBox="0 0 1500 350"
            width="1500"
          >
            <defs>
              <clippath
                id="recharts1-clip"
              >
                <rect
                  height="310"
                  width="1430"
                  x="65"
                  y="5"
                />
              </clippath>
            </defs>
            <g
              class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
            >
              <line
                class="recharts-cartesian-axis-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                type="category"
                width="1430"
                x="65"
                x1="65"
                x2="1495"
                y="315"
                y1="315"
                y2="315"
              />
              <g
                class="recharts-cartesian-axis-ticks"
              >
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    type="category"
                    width="1430"
                    x="65"
                    x1="422.5"
                    x2="422.5"
                    y="315"
                    y1="321"
                    y2="315"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    type="category"
                    width="1430"
                    x="422.5"
                    y="323"
                  >
                    <tspan
                      dy="0.71em"
                      x="422.5"
                    >
                      0
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    type="category"
                    width="1430"
                    x="65"
                    x1="1137.5"
                    x2="1137.5"
                    y="315"
                    y1="321"
                    y2="315"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    type="category"
                    width="1430"
                    x="1137.5"
                    y="323"
                  >
                    <tspan
                      dy="0.71em"
                      x="1137.5"
                    >
                      auto
                    </tspan>
                  </text>
                </g>
              </g>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
            >
              <line
                class="recharts-cartesian-axis-line"
                fill="none"
                height="310"
                orientation="left"
                stroke="#666"
                type="number"
                width="60"
                x="5"
                x1="65"
                x2="65"
                y="5"
                y1="5"
                y2="315"
              />
              <g
                class="recharts-cartesian-axis-ticks"
              />
            </g>
          </svg>
          <div
            class="recharts-legend-wrapper"
            style="position: absolute; width: 1490px; height: auto; left: 5px; bottom: 5px;"
          >
            <ul
              class="recharts-default-legend"
              style="padding: 0px; margin: 0px; text-align: center;"
            >
              <li
                class="recharts-legend-item legend-item-0"
                style="display: inline-block; margin-right: 10px;"
              >
                <svg
                  class="recharts-surface"
                  height="14"
                  style="display: inline-block; vertical-align: middle; margin-right: 4px;"
                  version="1.1"
                  viewBox="0 0 32 32"
                  width="14"
                >
                  <path
                    class="recharts-legend-icon"
                    d="M0,4h32v24h-32z"
                    fill="#2976d2"
                    stroke="none"
                  />
                </svg>
                <span
                  class="recharts-legend-item-text"
                  style="color: rgb(41, 118, 210);"
                >
                  days
                </span>
              </li>
            </ul>
          </div>
          <div
            class="recharts-tooltip-wrapper"
            style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
          >
            <div
              class="recharts-default-tooltip"
              style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
            >
              <p
                class="recharts-tooltip-label"
                style="margin: 0px;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-8 css-am41oj-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
        >
          <table
            aria-label="Leave details"
            class="MuiTable-root css-dwyv7x-MuiTable-root"
          >
            <thead
              class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  Leave Type
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  No. Of Days
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  From
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  To
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  Status
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            />
          </table>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-4 css-1ube6tp-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
        >
          <table
            aria-label="Leave balance"
            class="MuiTable-root css-dwyv7x-MuiTable-root"
          >
            <thead
              class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
                  scope="col"
                >
                  Leave Type
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
                  scope="col"
                >
                  Balance
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            />
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
